### PR TITLE
30 add database entites

### LINKED ISSUE #30: Add database entities

### PR DESCRIPTION
**Pull request by Skrabka98**
Closes #30

### CODE REVIEW

File: TribalWars/src/ApplicationCore/Enums/BattleSide.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ApplicationCore.Enums
+{
+    public enum BattleSide
+    {
+        attacker,
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Enums/BattleSide.cs]: Enum values should be named in PascalCase.

File: TribalWars/src/ApplicationCore/Enums/TribeRole.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ApplicationCore.Enums
+{
+    public enum TribeRole
+    {
+        owner,
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Enums/TribeRole.cs]: Enum values should be named in PascalCase.

File: TribalWars/src/ApplicationCore/Entities/ArmyUnitRequiredMaterial.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class ArmyUnitRequiredMaterial
+    {
+        public int Id { get; set; }
+        public int ArmyUnitTypeId { get; set; }
+        public int MaterialId { get; set; }
+        public int Quantity { get; set; }
+        public virtual ArmyUnitType ArmyUnitType { get; set; }
+        public virtual Material Material { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/ArmyUnitRequiredMaterial.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/Building.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        public int ConstructionTime { get; set; }
+        public int MaxLevel { get; set; }
+        public string Type { get; set; }
+        public int BaseValue { get; set; }
+        public virtual ICollection<ArmyUnitType> ArmyUnitTypes { get; set; }
+        public virtual ICollection<BuildingQueue> BuildingQueues { get; set; }
+        public virtual ICollection<BuildingRequiredBuilding> BuildingRequiredBuildings { get; set; }
+        public virtual ICollection<VillageBuilding> VillageBuildings { get; set; }
+        public virtual ICollection<BuildingRequiredMaterial> BuildingRequiredMaterials { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/Building.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/BuildingQueue.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    public class BuildingQueue
+    {
+        public int Id { get; set; }
+        public int VillageId { get; set; }
+        public int BuildingId { get; set; }
+        public DateTime StartDate { get; set; }
+        public int Duration { get; set; }
+        public virtual Village Village { get; set; }
+        public virtual Building Building { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/BuildingQueue.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+namespace ApplicationCore.Entities
+{
+    public class BuildingRequiredBuilding
+    {
+        public int Id { get; set; }
+        public int BuildingId { get; set; }
+        public int RequiredBuildingId { get; set; }
+        public int Level { get; set; }
+        public virtual Building Building { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/BuildingRequiredMaterial.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public class BuildingRequiredMaterial
+    {
+        public int Id { get; set; }
+        public int MaterialId { get; set; }
+        public int BuildingId { get; set; }
+        public int Quantity { get; set; }
+        public virtual Building Building { get; set; }
+        public virtual Material Material { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/BuildingRequiredMaterial.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/Material.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class Material
+    {
+        public int Id { get; set; }
+        public string Name { get; set; }
+        public virtual ICollection<ArmyUnitRequiredMaterial> ArmyUnitRequiredMaterials { get; set; }
+        public virtual ICollection<BuildingRequiredMaterial> BuildingRequiredMaterials { get; set; }
+        public virtual ICollection<StolenMaterial> StolenMaterials { get; set; }
+        public virtual ICollection<VillageMaterial> VillageMaterials { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/Material.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/Player.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class Player
+    {
+        public Guid Id { get; set; }
+        public Guid UserId { get; set; }
+        public string Nickname { get; set; }
+        public DateTime CreatedAt { get; set; }
+        public virtual Village Village { get; set; }
+        public ICollection<TribePlayer> TribePlayer { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/Player.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/RecruitmentQueue.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public int Id { get; set; }
+        public int VillageId { get; set; }
+        public int ArmyUnitTypeId { get; set; }
+        public int Quantity { get; set; }
+        public DateTime StartData { get; set; }
+        public int Duration { get; set; }
+        public virtual Village Village { get; set; }
+        public virtual ArmyUnitType ArmyUnitType { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/RecruitmentQueue.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/StolenMaterial.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class StolenMaterial
+    {
+        public int Id { get; set; }
+        public int MaterialId { get; set; }
+        public int AttackId { get; set; }
+        public int Quantity { get; set; }
+        public virtual Attack Attack { get; set; }
+        public virtual Material Material { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/StolenMaterial.cs]: Useless whitelane. 3 times!!

File: TribalWars/src/ApplicationCore/Entities/VillageBuilding.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public class VillageBuilding
+    {
+        public int Id { get; set; }
+        public int VillageId { get; set; }
+        public int BuildingId { get; set; }
+        public int CurrentLevel { get; set; }
+        public virtual Village Village { get; set; }
+        public virtual Building Building { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/VillageBuilding.cs]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/VillageStatistic.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class VillageStatistic
+    {
+        public int Id { get; set; }
+        public int MaterialsCapacity { get; set; }
+        public int PeopleCapacity { get; set; }
+        public int WallDefence { get; set; }
+        public int VillageId { get; set; }
+        public virtual Village Village { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/VillageStatistic.cs]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/AttackConfig.cs [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+using System.Threading.Tasks;
+
+namespace Infrastructure.Configuration
+{
+    public class AttackConfig : IEntityTypeConfiguration<Attack>
+    {
+        public void Configure(EntityTypeBuilder<Attack> builder)
+        {
+            builder.HasKey(i => i.Id);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/AttackConfig.cs]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/AttackReportConfig.cs [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    class AttackReportConfig : IEntityTypeConfiguration<AttackReport>
+    {
+        public void Configure(EntityTypeBuilder<AttackReport> builder)
+        {
+            builder.HasKey(i => i.Id);
+
+            builder.HasOne<Attack>(a => a.Attack)
+                .WithOne(x => x.AttackReport)
+                .HasForeignKey<AttackReport>(a => a.AttackId);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/AttackReportConfig.cs:17]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/AttackReportConfig.cs [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        }
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/AttackReportConfig.cs:22]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/BattleUnitConfig.cs [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+               .HasOne<ArmyUnitType>(a => a.ArmyUnitType)
+               .WithMany(x => x.BattleUnits)
+               .HasForeignKey(a => a.ArmyUnitTypeId);
+
+            builder
+                .Property(x => x.Side)
+                .HasConversion
+                    (v => v.ToString(),
+                    v => (BattleSide)Enum.Parse(typeof(BattleSide), v));
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/BattleUnitConfig.cs]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/BuildingConfig.cs [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Infrastructure.Configuration
+{
+    public class BuildingConfig : IEntityTypeConfiguration<Building>
+    {
+        public void Configure(EntityTypeBuilder<Building> builder)
+        {
+            builder
+                .HasKey(i => i.Id);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/BuildingConfig.cs]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/BuildingQueueConfig.cs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+            builder
+                .HasOne<Building>(b => b.Building)
+                .WithMany(x => x.BuildingQueues)
+                .HasForeignKey(b => b.BuildingId);
+
+            builder
+               .HasOne<Village>(v => v.Village)
+               .WithMany(x => x.BuildingQueues)
+               .HasForeignKey(v => v.VillageId);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/BuildingQueueConfig.cs]: Useless whitespace.

File: TribalWars/src/ApplicationCore/Entities/VillageMaterial.cs [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    public class VillageMaterial
+    {
+        public int Id { get; set; }
+        public int VillageId { get; set; }
+        public int MaterialId { get; set; }
+        public int Quantity { get; set; }
+        public float PerMinute { get; set; }
+        public virtual Village Village { get; set; }
+        public virtual Material Material { get; set; }
+
```
Comment (wikhag) [inline TribalWars/src/ApplicationCore/Entities/VillageMaterial.cs]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/BuildingRequiredMaterialConfig.cs [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+            builder.HasKey(i => i.Id);
+
+            builder.HasOne<Building>(b => b.Building)
+                .WithMany(x => x.BuildingRequiredMaterials)
+                .HasForeignKey(b => b.BuildingId);
+
+            builder.HasOne<Material>(m => m.Material)
+               .WithMany(x => x.BuildingRequiredMaterials)
+               .HasForeignKey(m => m.MaterialId);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/BuildingRequiredMaterialConfig.cs:17]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/VillageConfig.cs [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    class VillageConfig : IEntityTypeConfiguration<Village>
+    {
+        public void Configure(EntityTypeBuilder<Village> builder)
+        {
+            builder.HasKey(i => i.Id);
+
+            builder.HasOne<Player>(p => p.Player)
+                .WithOne(x => x.Village)
+                .HasForeignKey<Village>(v => v.PlayerId);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/VillageConfig.cs:17]: Useless whitelane.

File: TribalWars/src/Infrastructure/Configuration/VillageMaterialConfig.cs [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+            builder.HasKey(i => i.Id);
+
+            builder.HasOne<Material>(m => m.Material)
+                .WithMany(x => x.VillageMaterials)
+                .HasForeignKey(m => m.MaterialId);
+
+            builder.HasOne<Village>(v => v.Village)
+                .WithMany(x => x.VillageMaterials)
+                .HasForeignKey(v => v.VillageId);
+
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Configuration/VillageMaterialConfig.cs:21]: Useless whitelane.

File: TribalWars/src/ApplicationCore/Entities/ArmyUnitRequiredMaterial.cs (hunk 3 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/ArmyUnitRequiredMaterial.cs]: empty line

File: TribalWars/src/ApplicationCore/Entities/Building.cs (hunk 4 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/Building.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/BuildingQueue.cs [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/BuildingQueue.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs (hunk 6 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/BuildingRequiredMaterial.cs [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/BuildingRequiredMaterial.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/Material.cs (hunk 8 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/Material.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/Player.cs (hunk 9 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/Player.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/RecruitmentQueue.cs (hunk 10 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/RecruitmentQueue.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/StolenMaterial.cs [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+
+
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/StolenMaterial.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/VillageBuilding.cs [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/VillageBuilding.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/VillageMaterial.cs [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/VillageMaterial.cs]: empty lines

File: TribalWars/src/ApplicationCore/Entities/VillageStatistic.cs (hunk 13 above)
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Entities/VillageStatistic.cs]: empty lines

File: TribalWars/src/ApplicationCore/Enums/BattleSide.cs [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+        defender
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Enums/BattleSide.cs]: the C# convention says that enums should also be written pascal case

File: TribalWars/src/ApplicationCore/Enums/TribeRole.cs [hunk 30]
```diff
@@ -0,0 +1,14 @@
...
+        member
```
Comment (bboysokol) [inline TribalWars/src/ApplicationCore/Enums/TribeRole.cs]: the C# convention says that enums should also be written pascal case

File: TribalWars/src/Infrastructure/Configuration/ArmyUnitRequiredMaterialsConfig.cs [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            builder.HasKey(i => i.Id);
+
+            builder.HasOne<ArmyUnitType>(a => a.ArmyUnitType)
+                .WithMany(x => x.ArmyUnitRequiredMaterials)
+                .HasForeignKey(x => x.ArmyUnitTypeId);
+
+            builder.HasOne<Material>(m => m.Material)
+                .WithMany(m => m.ArmyUnitRequiredMaterials)
+                .HasForeignKey(x => x.MaterialId);
```
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/ArmyUnitRequiredMaterialsConfig.cs]: for information only:
for better understanding for the future, better way of naming variables in arrow function is to use meaningful var, which describe objects which we have inside
e.g. 
```
 builder.HasOne<ArmyUnitType>(requirements => requirements.ArmyUnitType)**
                .WithMany(unitType => unitType.ArmyUnitRequiredMaterials)
                .HasForeignKey(requirement => requirement.ArmyUnitTypeId)
```

Comment (bboysokol) [reply TribalWars/src/Infrastructure/Configuration/ArmyUnitRequiredMaterialsConfig.cs]: we use this approach to better understand what the action is about

File: TribalWars/src/Infrastructure/Configuration/AttackConfig.cs (hunk 14 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/AttackConfig.cs]: empty line

File: TribalWars/src/Infrastructure/Configuration/AttackReportConfig.cs (hunk 16 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/AttackReportConfig.cs]: empty lines

File: TribalWars/src/Infrastructure/Configuration/ArmyUnitRequiredMaterialsConfig.cs (hunk 31 above)
Comment (bboysokol) [reply TribalWars/src/Infrastructure/Configuration/ArmyUnitRequiredMaterialsConfig.cs]: I leave you to consider whether you will change it or not

File: TribalWars/src/Infrastructure/Configuration/BattleUnitConfig.cs (hunk 17 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/BattleUnitConfig.cs]: empty line

File: TribalWars/src/Infrastructure/Configuration/BuildingConfig.cs (hunk 18 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/BuildingConfig.cs]: empty line

File: TribalWars/src/Infrastructure/Configuration/BuildingQueueConfig.cs (hunk 19 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/BuildingQueueConfig.cs]: empty line

File: TribalWars/src/Infrastructure/Configuration/BuildingRequiredMaterialConfig.cs [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+
```
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/BuildingRequiredMaterialConfig.cs]: empty lines

File: TribalWars/src/Infrastructure/Configuration/VillageBuildingConfig.cs [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+            builder.HasKey(i => i.Id);
+
+            builder.HasOne<Village>(v => v.Village)
+                .WithMany(x => x.VillageBuildings)
+                .HasForeignKey(v => v.VillageId);
+
+            builder.HasOne<Building>(b => b.Building)
+                .WithMany(x => x.VillageBuildings)
+                .HasForeignKey(b => b.BuildingId);
+
```
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/VillageBuildingConfig.cs:21]: empty lines

File: TribalWars/src/Infrastructure/Configuration/VillageConfig.cs (hunk 22 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/VillageConfig.cs:17]: empty lines

File: TribalWars/src/Infrastructure/Configuration/VillageMaterialConfig.cs (hunk 23 above)
Comment (bboysokol) [inline TribalWars/src/Infrastructure/Configuration/VillageMaterialConfig.cs:21]: empty lines

**Review by bboysokol — CHANGES_REQUESTED** [review]

File: TribalWars/src/Infrastructure/Migrations/20210430085658_add_attackVillage.Designer.cs [hunk 34]
```diff
@@ -0,0 +1,1158 @@
...
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Migrations;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace Infrastructure.Migrations
+{
+    [DbContext(typeof(ApplicationDbContext))]
+    [Migration("20210430085658_add_attackVillage")]
```
Comment (wikhag) [inline TribalWars/src/Infrastructure/Migrations/20210430085658_add_attackVillage.Designer.cs]: Try to have consistency in migration naming. For example - don't mix snake_case with camelCase. Name this migration like this 'add_attack_village'.
You have to change Migration() attribute and file name.

**Pull request merged**